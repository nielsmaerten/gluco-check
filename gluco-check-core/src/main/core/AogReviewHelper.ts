import {logger, config} from 'firebase-functions';
import {injectable} from 'inversify';
import User from '../../types/User';
import FirebaseClient from '../clients/FirebaseClient';
import {sampleData, aogAgentRegex} from '../constants';
const logTag = '[NewUserHook]';

@injectable()
export default class AogReviewHelper {
  constructor(private firebaseClient: FirebaseClient) {}

  /**
   * Creates/Updates profile of the specified user with sample data
   * This is used to provide sample data to Google Action reviewers
   */
  async addSampleData(user: User) {
    // Create reference to user document in Firestore
    const userId = user.userId;
    const userDocumentPath = `users/${userId}`;
    const userDocumentRef = this.firebaseClient.firestore().doc(userDocumentPath);

    // Build user document with sample data
    const userDocument: User = {
      defaultMetrics: sampleData.defaultMetrics,
      glucoseUnit: sampleData.glucoseUnit,
      nightscout: config().nightscout_for_testers,
      exists: true,
      userId: userId,
      heardDisclaimer: false,
    };

    logger.warn(logTag, `Populating test user ${userId} with sample data`);
    await userDocumentRef.set(userDocument, {merge: true});
    Object.assign(user, userDocument);
  }

  /**
   * Returns TRUE if the email address matches that of an AOG platform reviewer
   * @param user The user's email address
   */
  isReviewer(user: User) {
    const email = user.userId;
    const isTestUser = aogAgentRegex.test(email);
    return isTestUser;
  }
}
