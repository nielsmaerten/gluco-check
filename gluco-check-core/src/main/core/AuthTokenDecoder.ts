/* istanbul ignore file: External system (Google Actions) */
import {ConversationV3} from '@assistant/conversation';
import {AuthHeaderProcessor} from '@assistant/conversation/dist/auth';
import {logger, config} from 'firebase-functions';
import {injectable} from 'inversify';

@injectable()
/**
 * Decodes the Authorization header of a request coming from Google Assistant
 * into a JWT with all user properties and adds this to the conversation object
 */
export default class AuthTokenDecoder {
  private clientId: string;
  private authHeaderProcessor = new AuthHeaderProcessor();

  constructor() {
    logger.debug('[AuthTokenDecoder]: Initializing new instance');
    this.clientId = config().google_actions_sdk.client_id;

    if (!this.clientId)
      throw '[AuthTokenDecoder]: Firebase Functions config must define an auth.client_id property';
  }

  async decodeGoogleUserToken(conversation: ConversationV3) {
    const header = '' + conversation.headers.authorization;
    const processor = this.authHeaderProcessor;
    const clientId = this.clientId;

    await conversation.user.processAuthHeader(header, processor, clientId);
  }
}
