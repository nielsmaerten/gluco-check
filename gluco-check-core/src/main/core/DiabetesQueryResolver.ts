import DiabetesQuery from '../../types/DiabetesQuery';
import AssistantResponse from '../../types/AssistantResponse';
import {ErrorTypes} from '../../types/ErrorTypes';
import DiabetesSnapshot from '../../types/DiabetesSnapshot';
import NightscoutClient from '../clients/NightscoutClient';
import {injectable} from 'inversify';
import ResponseFormatter from './ResponseFormatter';
import {logger} from 'firebase-functions';
import {performance} from 'perf_hooks';

/**
 * DiabetesQuery resolver accepts a DiabetesSnapshot and turns it into an AssistantResponse.
 * It is responsible for getting the data for each Pointer in the Query by calling Clients
 */
@injectable()
export default class DiabetesQueryResolver {
  constructor(private responseFormatter: ResponseFormatter) {}

  async resolve(query: DiabetesQuery): Promise<AssistantResponse> {
    // Ensure user exists and has a Nightscout Site
    if (!query.user.exists || !query.user.nightscout) {
      return this.responseFormatter.formatError(ErrorTypes.UserNotFound, query);
    }

    // Build a new snapshot and Client
    const snapshot = new DiabetesSnapshot(Date.now());
    snapshot.glucoseUnit = query.user.glucoseUnit;
    const nsClient = new NightscoutClient(query.user.nightscout);

    // Perform queries for each pointer
    const queryStart = performance.now();
    const snapshotParts = await Promise.all(
      query.pointers.map(p => nsClient.getPointer(p))
    );
    const totalQueryTime = Math.floor(performance.now() - queryStart);

    // Merge partial snapshots together
    snapshotParts.forEach(part => {
      Object.assign(snapshot, part);
    });

    logger.info(`[DiabetesQueryResolver]: Total query time: ${totalQueryTime} ms.`);
    return this.responseFormatter.formatSnapshot(snapshot, query);
  }
}
