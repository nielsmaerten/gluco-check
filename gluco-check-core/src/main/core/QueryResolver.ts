import DmQuery from '../../types/DmQuery';
import {ErrorType} from '../../types/ErrorType';
import DmSnapshot from '../../types/DmSnapshot';
import NightscoutClient from '../clients/nightscout/NightscoutClient';
import {injectable} from 'inversify';
import {logger} from 'firebase-functions';
import {performance} from 'perf_hooks';
import {DmMetric} from '../../types/DmMetric';

@injectable()
export default class QueryResolver {
  /**
   * Resolves a DmQuery by fetching all requested metrics from Nightscout.
   * Returns a DmSnapshot with all of the fetched values
   */
  async buildSnapshot(query: DmQuery): Promise<DmSnapshot> {
    // Ensure user exists and has a Nightscout Site
    if (!query.user.exists || !query.user.nightscout) {
      return this.userNotFoundError(query);
    }

    // Create an empty snapshot. We'll add requested Metrics in here
    const newSnapshot = new DmSnapshot({
      query: query,
      timestamp: Date.now(),
      glucoseUnit: query.user.glucoseUnit,
    });

    // Create a Nightscout Client we can use for querying
    const nsClient = new NightscoutClient(query.user.nightscout);

    // Query Nightscout for each requested metric
    const queryStart = performance.now();
    const promises = query.metrics.map(p => nsClient.getMetric(p) as Partial<DmSnapshot>);
    const partialSnapshots = await Promise.all(promises);
    const totalQueryTime = Math.floor(performance.now() - queryStart);
    logger.info(`[DiabetesQueryResolver]: Nightscout queries took ${totalQueryTime} ms.`);

    // Add all partial snapshots into the main snapshot
    newSnapshot.update(...partialSnapshots);
    return newSnapshot;
  }

  private userNotFoundError(query: DmQuery) {
    return new DmSnapshot({
      timestamp: Date.now(),
      query,
      errors: [
        {
          type: ErrorType.Firebase_UserNotFound,
          affectedMetric: DmMetric.Everything,
        },
      ],
    });
  }
}
