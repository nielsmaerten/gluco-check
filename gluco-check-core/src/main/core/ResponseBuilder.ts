import {injectable} from 'inversify';

import InternationalizationHelper from '../i18n';
import DmSnapshot from '../../types/DmSnapshot';
import AssistantResponse from '../../types/AssistantResponse';

import {ErrorType} from '../../types/ErrorType';
import {DmMetric} from '../../types/DmMetric';
import Humanizer from '../i18n/humanizers';
import DmQuery from '../../types/DmQuery';
import UserProfileClient from '../clients/UserProfileClient';

@injectable()
export default class ResponseBuilder {
  constructor(
    private localizer: InternationalizationHelper,
    private userClient: UserProfileClient
  ) {}

  async build(snapshot: DmSnapshot): Promise<AssistantResponse> {
    // Ensure the required locale is available
    await this.localizer.ensureLocale(snapshot.query.locale);

    // Handle general errors, if any
    const generalError = findGeneralErrors(snapshot);
    if (generalError) return this.errorResponse(snapshot, generalError);
    // Build a normal response
    else return this.normalResponse(snapshot);
  }

  private async normalResponse(snapshot: DmSnapshot): Promise<AssistantResponse> {
    // Humanize every metric in the Snapshot
    const humanizedMetrics: string[] = await Humanizer.dmSnapshot(snapshot);

    // Wrap every humanized string in an <s>-tag
    // Note the space at the end for readability --------â–½
    const s_tags = humanizedMetrics.map(txt => `<s>${txt} </s>`);

    // Join all tags together to create the SSML string
    const output = s_tags.join('');
    const disclaimer = medicalDisclaimer(snapshot.query);
    const SSML = `<speak>${output + disclaimer}</speak>`;

    // If disclaimer will be mentioned, mark it as heard
    const userHeardDisclaimer = disclaimer !== '';
    const user = snapshot.query.user;
    await this.userClient.flagDisclaimer(user, userHeardDisclaimer, false);

    return new AssistantResponse(SSML, snapshot.query.locale);
  }

  private errorResponse(
    snapshot: DmSnapshot,
    error: {type: ErrorType; affectedMetric: DmMetric}
  ) {
    const errorTxt = Humanizer.error(error.type, snapshot.query.locale);
    const disclaimer = medicalDisclaimer(snapshot.query);
    const SSML = `<speak>${errorTxt + disclaimer}</speak>`;
    return new AssistantResponse(SSML, snapshot.query.locale);
  }
}
/**
 * General errors are situations that cause the entire query to fail.
 * None of the requested metrics will be resolved.
 */
function findGeneralErrors(snapshot: DmSnapshot) {
  const generalErrors = [
    ErrorType.Nightscout_Unauthorized,
    ErrorType.Nightscout_Unavailable,
    ErrorType.Firebase_UserNotFound,
  ];
  return snapshot.errors.find(e => generalErrors.includes(e.type));
}

function medicalDisclaimer(query: DmQuery) {
  if (!query.metadata.mentionDisclaimer) {
    return '';
  } else {
    return Humanizer.disclaimer(query.locale);
  }
}
