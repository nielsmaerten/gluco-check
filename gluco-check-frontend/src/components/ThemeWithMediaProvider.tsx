import React, { ReactNode } from "react";
import {
  createMuiTheme,
  responsiveFontSizes,
  ThemeOptions,
  ThemeProvider,
  useMediaQuery,
} from "@material-ui/core";

type ThemeWithMediaProviderProps = {
  themeOptions: ThemeOptions;
  children: ReactNode;
};

// Component for mixing user media preferences into MUI theme,
// currently used to add dark mode support
function ThemeWithMediaProvider({
  children,
  themeOptions,
}: ThemeWithMediaProviderProps) {
  const prefersDark = useMediaQuery("(prefers-color-scheme: dark)");

  const theme = React.useMemo(() => {
    let blendedTheme = createMuiTheme({
      ...themeOptions,
      palette: {
        type: prefersDark ? "dark" : "light",
      },
    });
    blendedTheme = responsiveFontSizes(blendedTheme);
    return blendedTheme;
  }, [prefersDark, themeOptions]);

  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
}

export default ThemeWithMediaProvider;
