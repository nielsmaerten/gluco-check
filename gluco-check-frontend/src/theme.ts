/* istanbul ignore file */
import red from "@material-ui/core/colors/red";
import orange from "@material-ui/core/colors/orange";
import { ThemeOptions } from "@material-ui/core";

export const themeOptions: ThemeOptions = {
  typography: {
    fontFamily: ["Poppins", "sans-serif"].join(","),
  },
  palette: {
    text: {
      primary: "#202124",
    },
    primary: {
      main: "#1a73e8",
      contrastText: "#fff",
    },
    secondary: {
      main: "#4285f4",
    },
    error: {
      main: red[500],
    },
    warning: {
      light: "rgba(245, 124, 0, 0.75)", // orange[700] at 75%
      main: orange[700],
    },
    background: {
      default: "#fff",
    },
  },
};
