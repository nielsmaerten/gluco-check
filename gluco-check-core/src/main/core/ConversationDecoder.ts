import {ConversationV3} from '@assistant/conversation';
import DmQuery from '../../types/DmQuery';
import {injectable} from 'inversify';
import {DmMetric} from '../../types/DmMetric';
import AuthTokenDecoder from './AuthTokenDecoder';
import UserProfileClient from '../clients/UserProfileClient';
import {logger, config} from 'firebase-functions';
import User from '../../types/User';

@injectable()
/**
 * The ConversationDecoder figures out what the user actually asked for.
 * It takes in a 'conversation' object from Google Actions, and decodes it into a DmQuery
 */
export default class ConversationDecoder {
  private lastKnownActionVersion: number;

  constructor(
    private authTokenDecoder: AuthTokenDecoder,
    private userProfileClient: UserProfileClient
  ) {
    this.lastKnownActionVersion = this.getLastKnownActionVersion();
  }

  /**
   * Interprets an Assistant Conversation object and returns a DmQuery
   * representing what the user asked for
   */
  async decode(conv: ConversationV3): Promise<DmQuery> {
    // Populate conversation.user with info from json web token
    await this.authTokenDecoder.decodeGoogleUserToken(conv);

    // The following info needs to be extracted from the Conversation:
    const locale = conv.user.locale;
    const userId = conv.user.params.tokenPayload.email;
    const user = await this.userProfileClient.getUser(userId);

    // Build DmQuery object with all info required to respond to the user
    const dmMetrics = await this.extractMetrics(conv, user);
    const dmQuery = new DmQuery(user, locale, dmMetrics);
    dmQuery.metadata.mentionDisclaimer = this.shouldMentionDisclaimer(conv, user);
    dmQuery.metadata.mentionMissingMetrics = this.isDeepInvocation(conv);

    // Log status
    logger.info(
      `[ConversationDecoder]: ${user.userId.substr(0, 7)}...`,
      `requested: ${dmQuery.metrics}`
    );
    if (user.exists) {
      logger.debug('[ConversationDecoder]: Processing query:', dmQuery);
    } else {
      logger.warn(
        `[ConversationDecoder]: '${userId}'`,
        'invoked Gluco Check but does not exist in db'
      );
    }
    return dmQuery;
  }

  /**
   * Disclaimer should be added to the response if:
   * - The user has not already heard it
   * - OR: Invoked using a newer version of the Action
   */
  private shouldMentionDisclaimer(conv: ConversationV3, user: User): boolean {
    // User hasn't heard disclaimer before
    if (!user.heardDisclaimer) return true;

    // Invoked using latest version of the Action
    const invokingActionVersion_raw = conv.headers['gluco-check-version'];
    const invokingActionVersion = parseInt(invokingActionVersion_raw as string);

    const usingNewerAction = invokingActionVersion > this.lastKnownActionVersion;
    if (usingNewerAction) {
      logger.info('Force mentioning disclaimer bc of newer Action calling');
    }
    return usingNewerAction;
  }

  /**
   * The last known version of the Google Action, is pulled from Firebase Config
   * You must set this value using the Firebase CLI before deploying
   *
   * Disclaimers will always be mentioned in the last known version
   */
  private getLastKnownActionVersion() {
    const versionString = config().google_actions_sdk.glucocheck_action_version;
    const version = parseInt(versionString);

    // Abort if the Action version has not been set
    if (!version) {
      const errMsg =
        '[ConversationDecoder]: Initialization failed: glucocheck_action_version must be set';
      logger.error(errMsg);
      throw new Error(errMsg);
    } else {
      logger.debug(
        '[ConversationDecoder]: Assuming',
        `v${version}`,
        'is the latest version of the Gluco Check Action being used.'
      );
      return version;
    }
  }

  /**
   * Extracts which DmMetrics were asked for in the conversation
   */
  private async extractMetrics(conv: ConversationV3, user: User): Promise<DmMetric[]> {
    const isDeepInvocation = this.isDeepInvocation(conv);
    const hasDmMetrics = conv.intent.params?.DmMetric !== undefined;

    if (isDeepInvocation && hasDmMetrics) {
      // Get requested metrics from intent params
      return conv.intent.params!.DmMetric!.resolved;
    } else {
      // Get requested metrics from user profile
      return user.defaultMetrics ?? [];
    }
  }

  private isDeepInvocation(conv: ConversationV3) {
    return conv.handler.name === 'custom_metrics';
  }
}
