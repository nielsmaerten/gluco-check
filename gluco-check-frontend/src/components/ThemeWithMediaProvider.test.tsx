import React from "react";
import { render } from "@testing-library/react";
import { themeOptions } from "../theme";
import ThemeWithMediaProvider from "./ThemeWithMediaProvider";

import { useMediaQuery, createMuiTheme } from "@material-ui/core";
jest.mock("@material-ui/core", () => {
  return {
    __esModule: true,
    ThemeProvider: (props: any) => {
      return <div>{props.children}</div>;
    },
    useMediaQuery: jest.fn().mockReturnValue(true),
    createMuiTheme: jest.fn(),
    responsiveFontSizes: jest.fn(),
  };
});

describe("ThemeWithMediaProvider", () => {
  it("renders the wrapped children", () => {
    const { container } = render(
      <ThemeWithMediaProvider themeOptions={themeOptions}>
        <div>Some Content</div>
      </ThemeWithMediaProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it("renders the appropriate theme when user prefers light", () => {
    expect.assertions(1);
    (useMediaQuery as jest.Mock).mockReturnValueOnce(false);
    (createMuiTheme as jest.Mock).mockImplementationOnce((theme) => {
      expect(theme.palette).toMatchSnapshot();
    });
    render(
      <ThemeWithMediaProvider themeOptions={themeOptions}>
        <div>Some Content</div>
      </ThemeWithMediaProvider>
    );
  });

  it("renders the appropriate theme when user prefers dark", () => {
    expect.assertions(1);
    (useMediaQuery as jest.Mock).mockReturnValueOnce(true);
    (createMuiTheme as jest.Mock).mockImplementationOnce((theme) => {
      expect(theme.palette).toMatchSnapshot();
    });
    render(
      <ThemeWithMediaProvider themeOptions={themeOptions}>
        <div>Some Content</div>
      </ThemeWithMediaProvider>
    );
  });
});
