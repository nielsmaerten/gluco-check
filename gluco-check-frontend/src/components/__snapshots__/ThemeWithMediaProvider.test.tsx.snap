// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeWithMediaProvider renders the appropriate theme when user prefers dark 1`] = `
Object {
  "type": "dark",
}
`;

exports[`ThemeWithMediaProvider renders the appropriate theme when user prefers light 1`] = `
Object {
  "type": "light",
}
`;

exports[`ThemeWithMediaProvider renders the wrapped children 1`] = `
<div>
  <div>
    <div>
      Some Content
    </div>
  </div>
</div>
`;
